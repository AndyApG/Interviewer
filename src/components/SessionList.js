import React from 'react';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { scrollable, selectable } from '../behaviours';
import { SessionCard } from '.';
import { entityPrimaryKeyProperty } from '../ducks/modules/network';
import {
  DragSource,
  MonitorDragSource,
} from '../behaviours/DragAndDrop';

const EnhancedSessionCard = DragSource(selectable(SessionCard));

/**
  * SessionCard List
  */
const SessionList = (props) => {
  const {
    className,
    details,
    progress,
    label,
    items,
    onItemClick,
    onDeleteCard,
    isItemSelected,
    getKey,
  } = props;

  const classNames = cx('session-list', className);

  return (
    <div className={classNames}>
      {
        items.map(session => (
          <EnhancedSessionCard
            key={getKey(session)}
            {...this.props}
            label={label(session)}
            selected={isItemSelected(session)}
            progress={progress(session)}
            details={details(session)}
            meta={() => ({ uuid: session.uuid })}
            onSelected={() => onItemClick(session)}
            onDeleteCard={() => onDeleteCard(session)}
          />
        ))
      }
    </div>
  );
};

SessionList.propTypes = {
  className: PropTypes.string,
  details: PropTypes.func,
  progress: PropTypes.func,
  label: PropTypes.func,
  items: PropTypes.array.isRequired,
  onItemClick: PropTypes.func,
  onDeleteCard: PropTypes.func,
  isItemSelected: PropTypes.func,
  getKey: PropTypes.func,
};

SessionList.defaultProps = {
  className: '',
  details: () => (''),
  progress: () => (''),
  label: () => (''),
  items: [],
  onItemClick: () => {},
  onDeleteCard: () => {},
  isItemSelected: () => false,
  itemType: 'SESSION',
  isDragging: false,
  meta: {},
  getKey: node => node[entityPrimaryKeyProperty],
};

export default compose(
  MonitorDragSource(['meta', 'isDragging']),
  scrollable,
)(SessionList);
