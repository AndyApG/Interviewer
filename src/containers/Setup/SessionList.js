import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { push } from 'react-router-redux';

import { actionCreators as sessionActions } from '../../ducks/modules/session';
import { CardList } from '../../components';

/**
  * Display stored sessions
  */
class SessionList extends Component {
  onClickLoadSession = (session) => {
    const pathname = this.props.getSessionPath(session.uid);
    this.props.setSession(session.uid);
    this.props.loadSession(pathname);
  }

  render() {
    const { sessions } = this.props;

    return (
      <CardList
        label={sessionInfo => sessionInfo.key}
        nodes={Object.keys(sessions).map(key => ({ uid: key, value: sessions[key] }))}
        onToggleCard={this.onClickLoadSession}
        details={sessionInfo => [
          { Path: sessionInfo.value.path },
          { Prompt: sessionInfo.value.promptIndex },
          { 'Number of Nodes': sessionInfo.value.network.nodes.length },
          { 'Number of Edges': sessionInfo.value.network.edges.length },
        ]}
      />
    );
  }
}

SessionList.propTypes = {
  getSessionPath: PropTypes.func.isRequired,
  loadSession: PropTypes.func.isRequired,
  sessions: PropTypes.object.isRequired,
  setSession: PropTypes.func.isRequired,
};

SessionList.defaultProps = {
};

function mapStateToProps(state) {
  return {
    getSessionPath: sessionId => state.sessions[sessionId].path,
    sessions: state.sessions,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    loadSession: path => dispatch(push(path)),
    setSession: bindActionCreators(sessionActions.setSession, dispatch),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(SessionList);
