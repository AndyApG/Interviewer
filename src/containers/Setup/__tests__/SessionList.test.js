/* eslint-env jest */

import React from 'react';
import { shallow, mount } from 'enzyme';
import { createStore } from 'redux';

import SessionList from '../SessionList';

const mockReduxState = {
  installedProtocols: {
    mockProtocol: {
      name: 'Protocol Name',
      codebook: {},
      assetManifest: {},
      description: '',
      forms: {},
      lastModified: '2018-10-01T00:00:00.000Z',
      stages: [],
    },
  },
  sessions: {
    a: {
      caseId: 'case1',
      network: { ego: {}, nodes: [0, 1, 2], edges: [0] },
      promptIndex: 2,
      protocolUID: 'mockProtocol',
      stageIndex: 2,
      updatedAt: 1554130548004,
    },
    b: {
      caseId: 'case22',
      network: { ego: {}, nodes: [], edges: [] },
      promptIndex: 1,
      protocolUID: 'mockProtocol',
      stageIndex: 2,
      updatedAt: 1554130540000,
    },
  },
  activeSessionId: 'case1',
};

describe('<SessionList />', () => {
  it('renders ok', () => {
    const component = shallow(<SessionList store={createStore(() => mockReduxState)} />);
    expect(component).toMatchSnapshot();
  });

  it('shows sessions as cards', () => {
    const component = mount(<SessionList store={createStore(() => mockReduxState)} />);
    expect(component.find('.card').length).toBe(2);
  });
});
