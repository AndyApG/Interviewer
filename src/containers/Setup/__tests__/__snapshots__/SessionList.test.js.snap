// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SessionList /> renders ok 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(SessionList)
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "getSessionPath": [Function],
      "loadSession": [Function],
      "protocolData": Object {},
      "removeSession": [Function],
      "sessions": Object {
        "a": Object {
          "name": "a name",
          "network": Object {
            "edges": Array [
              0,
            ],
            "nodes": Array [
              0,
              1,
              2,
            ],
          },
          "path": "/path/to/a",
          "promptIndex": 2,
          "protocolPath": "p",
          "updatedAt": 1528213062793,
        },
        "b": Object {
          "name": "b name",
          "network": Object {
            "edges": Array [],
            "nodes": Array [],
          },
          "path": "/path/to/b",
          "promptIndex": 1,
          "protocolPath": "p",
          "updatedAt": 1528218451710,
        },
      },
      "setSession": [Function],
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "getSessionPath": [Function],
        "loadSession": [Function],
        "protocolData": Object {},
        "removeSession": [Function],
        "sessions": Object {
          "a": Object {
            "name": "a name",
            "network": Object {
              "edges": Array [
                0,
              ],
              "nodes": Array [
                0,
                1,
                2,
              ],
            },
            "path": "/path/to/a",
            "promptIndex": 2,
            "protocolPath": "p",
            "updatedAt": 1528213062793,
          },
          "b": Object {
            "name": "b name",
            "network": Object {
              "edges": Array [],
              "nodes": Array [],
            },
            "path": "/path/to/b",
            "promptIndex": 1,
            "protocolPath": "p",
            "updatedAt": 1528218451710,
          },
        },
        "setSession": [Function],
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
